packages/formspree-core/src/submission.ts

```diff
@@ -0,0 +1,181 @@
+  readonly kind = 'error';
+
+  private readonly formErrors: FormError[] = [];
+  private readonly fieldErrors: Map<keyof T, FieldError[]> = new Map();
+
+  constructor(...serverErrors: ServerError[]) {
+    for (const err of serverErrors) {
+      // form errors
+      if (!err.field) {
+        this.formErrors.push({
+          code:
+            err.code && isFormErrorCode(err.code) ? err.code : 'UNSPECIFIED',
+          message: err.message,
+        });
+        continue;
+      }
+
+      const fieldErrors = this.fieldErrors.get(err.field) ?? [];
+      fieldErrors.push({
+        code: err.code && isFieldErrorCode(err.code) ? err.code : 'UNSPECIFIED',
+        message: err.message,
+      });
+      this.fieldErrors.set(err.field, fieldErrors);
+    }
+  }
+
+  getFormErrors(): readonly FormError[] {
+    return [...this.formErrors];
+  }
+
+  getFieldErrors<K extends keyof T>(field: K): readonly FieldError[] {
+    return this.fieldErrors.get(field) ?? [];
+  }
+
+  getAllFieldErrors(): readonly [keyof T, readonly FieldError[]][] {
+    return Array.from(this.fieldErrors);
+  }
+}
+
+export type FormError = {
+  readonly code: FormErrorCode | 'UNSPECIFIED';
+  readonly message: string;
+};
+
+function isFormErrorCode(code: string): code is FormErrorCode {
+  return code in FormErrorCodeEnum;
+}
+
+export type FormErrorCode = ValueOf<typeof FormErrorCodeEnum>;
+
+export const FormErrorCodeEnum = {
+  BLOCKED: 'BLOCKED',
+  EMPTY: 'EMPTY',
+  FILES_TOO_BIG: 'FILES_TOO_BIG',
+  FORM_NOT_FOUND: 'FORM_NOT_FOUND',
+  INACTIVE: 'INACTIVE',
+  NO_FILE_UPLOADS: 'NO_FILE_UPLOADS',
+  PROJECT_NOT_FOUND: 'PROJECT_NOT_FOUND',
+  TOO_MANY_FILES: 'TOO_MANY_FILES',
+  // Stripe
+  STRIPE_CLIENT_ERROR: 'STRIPE_CLIENT_ERROR',
+  STRIPE_SCA_ERROR: 'STRIPE_SCA_ERROR',
+} as const;
+
+export type FieldError = {
+  readonly code: FieldErrorCode | 'UNSPECIFIED';
+  readonly message: string;
+};
+
+function isFieldErrorCode(code: string): code is FieldErrorCode {
+  return code in FieldErrorCodeEnum;
+}
+
+export type FieldErrorCode = ValueOf<typeof FieldErrorCodeEnum>;
+
+export const FieldErrorCodeEnum = {
+  REQUIRED_FIELD_EMPTY: 'REQUIRED_FIELD_EMPTY',
+  REQUIRED_FIELD_MISSING: 'REQUIRED_FIELD_MISSING',
+  TYPE_EMAIL: 'TYPE_EMAIL',
+  TYPE_NUMERIC: 'TYPE_NUMERIC',
+  TYPE_TEXT: 'TYPE_TEXT',
+  // Stripe
+  STRIPE_CLIENT_ERROR: 'STRIPE_CLIENT_ERROR',
+  STRIPE_SCA_ERROR: 'STRIPE_SCA_ERROR',
+} as const;
+
+export function isServerErrorResponse(
+  obj: UnknownObject
+): obj is ServerErrorResponse {
+  return (
+    ('errors' in obj &&
+      Array.isArray(obj.errors) &&
+      obj.errors.every((err) => typeof err.message === 'string')) ||
+    ('error' in obj && typeof obj.error === 'string')
+  );
+}
+
+export type ServerErrorResponse = {
+  error: string;
+  errors?: ServerError[];
+};
+
+type ServerError = {
+  code?: string;
+  details?: Record<string, string>;
+  field?: string;
+  message: string;
+};
```

Comment:
Yeah, good question. It's just a type describing what we expect the server to send in case of an error. This matches the current implementation on the server but we only do runtime validation on the message (see https://github.com/formspree/formspree-js/pull/47/files#diff-1f9db01dd19d42e313a63f565d5ea2998a592f709e20fa1ec56f4a816f57ebdaR162-R164). As a result, they can drift and the type can be inaccurate if the future. It's an improvement than what we had though.